presto-prometheus/src/main/java/io/prestosql/plugin/prometheus/PrometheusSplitManager.java

```diff
@@ -0,0 +1,162 @@
+    {
+        PrometheusTableHandle tableHandle = (PrometheusTableHandle) connectorTableHandle;
+        PrometheusTable table = prometheusClient.getTable(tableHandle.getSchemaName(), tableHandle.getTableName());
+
+        // this can happen if table is removed during a query
+        if (table == null) {
+            throw new TableNotFoundException(tableHandle.toSchemaTableName());
+        }
+        List<ConnectorSplit> splits = new ArrayList<>();
+        generateTimesForSplits().forEach(time -> {
+            try {
+                splits.add(new PrometheusSplit(buildQuery(prometheusClient.config.getPrometheusURI(), time, table.getName())));
+            }
+            catch (URISyntaxException e) {
+                throw new UnsupportedOperationException(e.getMessage());
+            }
+        });
+        return new FixedSplitSource(splits);
+    }
+
+    // URIBuilder handles URI encode
+    private URI buildQuery(URI baseURI, String time, String tableName)
+            throws URISyntaxException
+    {
+        List<NameValuePair> nameValuePairs = new ArrayList<>(2);
+        nameValuePairs.add(new BasicNameValuePair("query",
+                tableName + "[" + prometheusClient.config.getQueryChunkSizeDuration() + "]"));
+        nameValuePairs.add(new BasicNameValuePair("time",
+                time));
+        return new URIBuilder(baseURI.toString())
+                .setPath("api/v1/query")
+                .setParameters(nameValuePairs).build();
+    }
+
+    /**
+     * Get the start times in seconds that divide up the query using the config value:
+     * maxQueryRangeDuration
+     * queryChunkSizeDuration
+     * The times will be used in queries to Prometheus like: `http://localhost:9090/api/v1/query?query=up[3w]&time=1568229904"`
+     * ** NOTE: Prometheus instant query wants the end time specified, you get duration before that time returned
+     *
+     * @return list of start times as epoch seconds, like ["1568053244", "1568926595"]
+     */
+    private List<String> generateTimesForSplits()
+    {
+        long maxQueryRangeDuration = timeUnitsToSeconds(prometheusClient.config.getMaxQueryRangeDuration());
+        long splitChunkSizeDuration = timeUnitsToSeconds(prometheusClient.config.getQueryChunkSizeDuration());
+        // if config `max-query-range-duration` less than `split-chunk-size-duration`, doesn't make sense, handle it
+        if (maxQueryRangeDuration < splitChunkSizeDuration) {
+            return ImmutableList.of();
+        }
+
+        int numChunks = (int) Math.round(maxQueryRangeDuration / splitChunkSizeDuration);
+        long firstEndTIme = (Instant.now().toEpochMilli() / MILLIS_IN_A_SECOND) - maxQueryRangeDuration + splitChunkSizeDuration;
+
+        ImmutableList.Builder<Long> builder = ImmutableList.builder();
+        IntStream.range(0, numChunks).forEachOrdered(n -> {
+            builder.add(firstEndTIme + n * splitChunkSizeDuration);
```

Comment:
Couldn't you skip the iteration at the bottom and just use a `ImmutableList.Builder<String>`:`builder.add(Long.toString(firstEndTIme + n * splitChunkSizeDuration))`. 

Also I feel like a builder shouldn't be necessary either, since you've already got a stream and could probably just collect the output. Can you map over `IntStream.range(0, numChunks)`?

Eg something like:
```
return IntStream.range(0, numChunks).map( n -> {
   return firstEndTIme + n * splitChunkSizeDuration
}).map(Long.toString).collect(Collectors.toList());
```

Reply:
Also i think this logic alone could be it's own method/function with it's own tests. I'm very curious to see the chunks.

Reply:
Nice catch, yeah, I think the return of that method is much cleaner now:
```        return IntStream.range(0, numChunks).mapToObj(n -> {
            long endTime = firstEndTime + n * splitChunkSizeDuration;
            if (endTime > nowSeconds) {
                endTime = nowSeconds;
            }
            return endTime;
        }).map(time -> Long.toString(time)).collect(Collectors.toList());``` in 2860d0a

Reply:
Well it does have several test against its output now, they test the resultant queries.